Hashtbl.create with negative initial guess

<!--
Welcome to OCaml's Issue tracker!

OCaml's developers use this tracker for bugs and feature requests only, rather
than user support.

If you have questions about *using* OCaml, please ask at
https://discuss.ocaml.org (more people read Discuss than this tracker, and
you'll get confirmation of whether you've really found a bug or need a new
feature).

If your error came from the OCaml package manager, opam, (messages beginning
`[ERROR] The compilation of ...`), please start at
https://github.com/ocaml/opam-repository/issues/new.

Some libraries and tools which used to be part of OCaml are now maintained
separately. Please post questions about Graphics, Num, camlp4, LablTk, CamlDBM
or OCamlTopWin on Discuss or on their respective issue trackers (see [README.adoc](https://github.com/ocaml/ocaml/blob/trunk/README.adoc#separately-maintained-components)
for a full list).
-->
The documentation of `Hashtbl.create` says 

> `Hashtbl.create n` creates a new, empty hash table, with initial size `n`. For best results, `n` should be on the order of the expected number of elements that will be in the table. The table grows as needed, so `n` is just an initial guess.

It is curious, that this function does not raise any exception when `n` is negative. It seems, that `Hashtbl.create (-10)` is the same as `Hashtbl.create 0`. While not a bug in itself, it does seem a bit strange at first, and there is no explanation in the documentation. 

We stumbled upon this while working on [ortac/qcheck-stm](https://github.com/ocaml-gospel/ortac), and I was encouraged by @jmid to report it here. (The Gospel contracts for the `Hashtbl` module can be found [here](https://github.com/ocaml-gospel/ortac/blob/0753c72567b83f50ff1d66342c12aafe7dd881b5/plugins/qcheck-stm/test/hashtbl.mli))

--- a/Changes
+++ b/Changes
@@ -79,6 +79,12 @@ Working version
 
 ### Manual and documentation:
 
+- #13469, #13474, #13535: Document that [Hashtbl.create n] creates a hash table
+  with a default minimal size, even if [n] is very small or negative.
+  (Antonin Décimo, Nick Bares, report by Nikolaus Huber and Jan Midtgaard,
+   review by Florian Angeletti, Anil Madhavapeddy, Gabriel Scherer,
+   and Miod Vallat)
+
 ### Compiler user-interface and warnings:
 
 - #13428: support dump=[source | parsetree | lambda | ... | cmm | ...]

--- a/stdlib/hashtbl.mli
+++ b/stdlib/hashtbl.mli
@@ -64,11 +64,12 @@ type (!'a, !'b) t
 
 val create : ?random: (* thwart tools/sync_stdlib_docs *) bool ->
              int -> ('a, 'b) t
-(** [Hashtbl.create n] creates a new, empty hash table, with
-   initial size [n].  For best results, [n] should be on the
-   order of the expected number of elements that will be in
-   the table.  The table grows as needed, so [n] is just an
-   initial guess.
+(** [Hashtbl.create n] creates a new, empty hash table, with initial
+   size greater or equal to the suggested size [n].  For best results,
+   [n] should be on the order of the expected number of elements that
+   will be in the table.  The table grows as needed, so [n] is just an
+   initial guess.  If [n] is very small or negative then it is
+   disregarded and a small default size is used.
 
    The optional [~random] parameter (a boolean) controls whether
    the internal organization of the hash table is randomized at each

--- a/stdlib/moreLabels.mli
+++ b/stdlib/moreLabels.mli
@@ -81,11 +81,12 @@ module Hashtbl : sig
 
   val create : ?random: (* thwart tools/sync_stdlib_docs *) bool ->
                int -> ('a, 'b) t
-  (** [Hashtbl.create n] creates a new, empty hash table, with
-     initial size [n].  For best results, [n] should be on the
-     order of the expected number of elements that will be in
-     the table.  The table grows as needed, so [n] is just an
-     initial guess.
+  (** [Hashtbl.create n] creates a new, empty hash table, with initial
+     size greater or equal to the suggested size [n].  For best results,
+     [n] should be on the order of the expected number of elements that
+     will be in the table.  The table grows as needed, so [n] is just an
+     initial guess.  If [n] is very small or negative then it is
+     disregarded and a small default size is used.
 
      The optional [~random] parameter (a boolean) controls whether
      the internal organization of the hash table is randomized at each

--- a/stdlib/templates/hashtbl.template.mli
+++ b/stdlib/templates/hashtbl.template.mli
@@ -64,11 +64,12 @@ type (!'a, !'b) t
 
 val create : ?random: (* thwart tools/sync_stdlib_docs *) bool ->
              int -> ('a, 'b) t
-(** [Hashtbl.create n] creates a new, empty hash table, with
-   initial size [n].  For best results, [n] should be on the
-   order of the expected number of elements that will be in
-   the table.  The table grows as needed, so [n] is just an
-   initial guess.
+(** [Hashtbl.create n] creates a new, empty hash table, with initial
+   size greater or equal to the suggested size [n].  For best results,
+   [n] should be on the order of the expected number of elements that
+   will be in the table.  The table grows as needed, so [n] is just an
+   initial guess.  If [n] is very small or negative then it is
+   disregarded and a small default size is used.
 
    The optional [~random] parameter (a boolean) controls whether
    the internal organization of the hash table is randomized at each

